### ISSUE TITLE
check_files.pl - Use of uninitialized value $shell_command

### ISSUE THREAD

**Issue by heinemannj** [willixix/naglio-plugins#27]
Status Information: Translated filespec 'dbq._' to regex 'dbq.._'
check_files.pl plugin version 0.41
Alarm at 25
Changed to directory '/var/spool/rsyslog'
**ePN /usr/lib/nagios/plugins/check_files.pl: "Use of uninitialized value $shell_command in concatenation (.) or string at (eval 1) line 619,".

---
## openSUSE 12.3

Service Information
Last Updated: Sat May 4 09:09:48 CEST 2013
Updated every 90 seconds
Nagios® Core™ 3.5.0 - www.nagios.org
Logged in as admin

---
# 'check_local_files' command definition
# 

define command{
 command_name   check_local_files
 command_line $USER1$/check_files.pl -a -f -L $ARG1$ -D $ARG2$ -F $ARG3$ -w $ARG4$ -c $ARG5$
}
# Define a service to check rsyslog queue.
# 

define service{
 use local-service
 hostgroup_name elma-server
 servicegroups elma-services
 service_description Syslog Queue
 check_command check_local_files!Syslog Queue!/var/spool/rsyslog!'dbq.*'!10!100
}

**Comment by willixix** [comment]
Ok, thanks for the report. Download latest version of check_files from github (I just committed it) and see if it works properly with it

**Comment by heinemannj** [comment]
```
Translated filespec 'dbq.*' to regex 'dbq..*'
```

check_files.pl plugin version 0.415
Alarm at 25
Changed to directory '/var/spool/rsyslog'
Command: LANG=C ls -l
Executing LANG=C ls -l 2>&1
got line: total 52
parsed: type=unset
got line: -rw------- 1 root root 446 Apr 21 19:00 dbq.00000004
**ePN /usr/lib/nagios/plugins/check_files.pl: "Use of uninitialized value $2 in pattern match (m//) at (eval 1) line 553, <SHELL_DATA> line 2.".

---

# 'check_local_files' command definition

#
define command{
 command_name   check_local_files
 command_line $USER1$/check_files.pl -v -a -f -D $ARG1$ -F $ARG2$ -w $ARG3$ -c $ARG4$
}

# Define a service to check rsyslog queue.

#
define service{
 use local-service
 hostgroup_name elma-server
 servicegroups elma-services
 service_description Syslog Queue
 check_command check_local_files!/var/spool/rsyslog!'dbq.*'!10!100
}

**Comment by willixix** [comment]
I'd really like to fix so this plugin works under ePN too, but how is this error even possible?
I have the following line there:
    if (defined($parsed[0]) && $parsed[0] =~ /([-d])(.{3})(.{3})(.{3})/) {
and then bunch of checks like
        $mod += 400 if $2 =~ /r/;
so $2 must be set if it matches initial regex. Why is it not?

**Comment by heinemannj** [comment]
Hi William,

I'am using the latest git repository (check_files.pl plugin version 0.415).
If I run the plugin from shell it works with no problems.
If I turn off Nagios embedded perl interprter it also works as expected.

So it seems that your plugin is missing some requirements of an ePN capable
plugin:

http://nagios.sourceforge.net/docs/3_0/epnplugins.html
http://docs.icinga.org/latest/en/epnplugins.html

My nagios.cfg -----------------------------------------------------------

# P1.PL FILE LOCATION

# This value determines where the p1.pl perl script (used by the

# embedded Perl interpreter) is located.  If you didn't compile

# Nagios with embedded Perl support, this option has no effect.

p1_file=/usr/lib/nagios/p1.pl

# EMBEDDED PERL INTERPRETER OPTION

# This option determines whether or not the embedded Perl interpreter

# will be enabled during runtime.  This option has no effect if Nagios

# has not been compiled with support for embedded Perl.

# Values: 0 = disable interpreter, 1 = enable interpreter

enable_embedded_perl=1

# EMBEDDED PERL USAGE OPTION

# This option determines whether or not Nagios will process Perl plugins

# and scripts with the embedded Perl interpreter if the plugins/scripts

# do not explicitly indicate whether or not it is okay to do so. Read

# the HTML documentation on the embedded Perl interpreter for more

# information on how this option works.

use_embedded_perl_implicitly=1

2013/5/4 William Leibzon notifications@github.com

**Comment by willixix** [comment]
Yeah, I know it works without ePN. And I know all about ePN requirements too. I made sure my plugins work with it and not only that but I've fixed several Perl libraries so that other plugins work too. I just don't understand why its failing here. The first issue you reported, I saw where the bug was and its fixed. But as far as last report, I don't get why its failing.

Now as far as ePN there are cases where plugins are just incompatible with it. A year ago I actually created a modified version of p1.pl to work around one of its issues I saw when using check_files:
  http://exchange.nagios.org/directory/Patches/Nagios-Core/nagios-embedded-perl-p1-2Epl-that-lets-redirect-STDOUT/details
You're welcome to try it but the issue you reported here is not related to that so I doubt that my modified p1.pl would make any difference.

I'll think about this more and if I have the time try to re-create this on my own.
